### ISSUE TITLE
Sphere / Cube data does not include system at centre

### ISSUE THREAD

**Issue by psdevil** [EDSM-NET/FrontEnd#460]
Hi EDSM Team,

I tried the following command in the browser and various other iterations of the call with Python (and even with Powershell).

```
https://www.edsm.net/api-v1/sphere-systems?systemname=Sol
```

Unfortunately it rarely seems to returns data on the Sol system itself, which I would have expected it to. This is not limited to just Sol but any system name or coordinates that I choose. Sometimes it does return the data for the selected system name, but mostly it doesn't. I'm not sure if this also happens with the cube equivalent as I've not tested it.

I have attached the latest dump I got for Sol from my browser and you will not find the name 'Sol' or a Distance of 0.

It's not the behaviour I'd expect from these calls, but I thought I ought to report it as a possible bug anyway. Feel free to correct me if I am wrong.

[sol-request-but-no-sol-data.txt](https://github.com/EDSM-NET/FrontEnd/files/7262021/sol-request-but-no-sol-data.txt)

**Comment by klightspeed** [comment]
Systems requests have a limit to the number of results they will return, and that limit will be exceeded in many places if no `radius` is specified.  Due to the way database indexes work, if less than half of the systems in the requested sphere are returned, then the results will not include the centre system.

**Comment by psdevil** [comment]
Perhaps you should make this clear in the documentation as this is not the implied behaviour.

From my reading of the documentation, and that of a couple of IT colleagues who have been testing my failing code, when you ask for all systems with a radius of 20ly, then you should get a list of all systems within 20ly of the centre. If you ask for the maximum of 100ly, then you should get all within 100ly etc.  There is no mention that you might only get a (potentially small) subset of the available systems.

We do appreciate that there have to be limits but, based on your documentation, we have assumed the issue was at our end. We have spent a couple of weeks trying to find out if the error was in the code, the local libraries, or even in the Python libraries.

So, from an end user perspective, I'm sure you can see why we think it's not doing what it seems it should.

Just to be clear, I do understand how SQL indexes work, and I always use a radius in my Python calls to the sphere API by using the request block below. I found it clearer in the code and more efficient in execution to always specify both the minimum and maximum radii, even when the minimum radius is almost always 0.
```
                # Create a request block for search by name
                requestData = {
                    "systemName":      systemName,
                    "radius":          radius,
                    "minRadius":       minRadius,
                    "showId":          1,
                    "showCoordinates": 1,
                    "showPermit":      1,
                    "showInformation": 1,
                    "showPrimaryStar": 1
                }
```

Could you please confirm whether or not there is any way to tell if you received all the systems within the radius or not, and if there is any way to receive the missing systems.

As always, please feel free to correct me if I am wrong about any of this.

**Comment by klightspeed** [comment]
I had a closer look at the data you posted.
The returned radius was 50ly, and there were 1238 systems, the 627th of which was:
```json
{"distance":0,"bodyCount":40,"name":"Sol"}
```

**Comment by AnthorNet** [comment]
I don't see any limit on the sphere api code.

**Comment by klightspeed** [comment]
> I don't see any limit on the sphere api code.

I stand corrected - <https://www.edsm.net/api-v1/sphere-systems?systemName=Sagittarius%20A%2A> returns 34154 systems in a 50ly radius, the 17187th of which is `Sagittarius A*` (though it does take a few seconds to return).

Trying to specify a radius larger than 100ly will get you `{}` (suggesting that Anthor is replacing empty arrays with empty objects to fix empty objects being encoded as arrays by PHP).  Perhaps it should return `400 Bad Request` instead?

**Comment by bgol** [comment]
> I have attached the latest dump I got for Sol from my browser and you will not find the name 'Sol' or a Distance of 0.
> [sol-request-but-no-sol-data.txt](https://github.com/EDSM-NET/FrontEnd/files/7262021/sol-request-but-no-sol-data.txt)

Bat there is at pos 34621: `{"distance":0,"bodyCount":40,"name":"Sol"}`

**Comment by psdevil** [comment]
**ISSUE SOLVED:**

---
I followed up on your above comments and can confirm that the expected system names **were** present in the EDSM API raw data streams returned from both the browser and in Python. This prompted further investigations.

The problem was found in the data returned from a 3rd party library we were using to bulk convert the JSON records to objects. It seems to have an undocumented built-in feature that silently removes what it thinks are duplicated data records based on the contents of the 'name', 'first' and 'last' fields. We only discovered this 'feature' when we tested the library with a variety of different JSON records outside of this project.

We have rewritten the library in our own Python code and although not as fast, the problem instantly vanished. The 3rd party library has been permanently removed.

---
Sorry to have bothered you - but thank you so much for the **_EXCELLENT_** and _**RAPID**_ support.
